### PR TITLE
[SPARK-23572][docs] Bring "security.md" up to date.

### PR DESCRIPTION
**Pull request by vanzin**
This change basically rewrites the security documentation so that it's
up to date with new features, more correct, and more complete.

Because security is such an important feature, I chose to move all the
relevant configuration documentation to the security page, instead of
having them peppered all over the place in the configuration page. This
allows an almost one-stop shop for security configuration in Spark. The
only exceptions are some YARN-specific minor features which I left in
the YARN page.

I also re-organized the page's topics, since they didn't make a lot of
sense. You had kerberos features described inside paragraphs talking
about UI access control, and other oddities. It should be easier now
to find information about specific Spark security features. I also
enabled TOCs for both the Security and YARN pages, since that makes it
easier to see what is covered.

I removed most of the comments from the SecurityManager javadoc since
they just replicated information in the security doc, with different
levels of out-of-dateness.

### CODE REVIEW

File: R/pkg/DESCRIPTION [hunk 1]
```diff
@@ -57,6 +57,6 @@ Collate:
     'types.R'
     'utils.R'
     'window.R'
-RoxygenNote: 5.0.1
+RoxygenNote: 6.0.1
```
Comment (felixcheung) [inline R/pkg/DESCRIPTION]: pls revert this

Comment (vanzin) [reply R/pkg/DESCRIPTION]: Argh, I though I had removed this one. (It's really annoying when the build changes your working file set.)

**Comment by vanzin** [comment]
Any takers for reviewing this?

**Comment by vanzin** [comment]
No volunteers, so I'll volunteer @squito , @jerryshao , @tgravescs

File: docs/security.md [hunk 2]
```diff
@@ -182,54 +580,70 @@ configure those ports.
...
+
+Long-running applications may run into issues if their run time exceeds the maximum delegation
+token lifetime configured in services it needs to access.
+
+Spark supports automatically creating new tokens for these applications when running in YARN mode.
+Kerberos credentials need to be provided to the Spark application via the `spark-submit` command,
+using the `--principal` and `--keytab` parameters.
+
+The provided keytab will be copied over to the machine running the Application Master via the Hadoop
+Distributed Cache. For this reason, it's strongly recommended that both YARN and HDFS encryption are
```
Comment (tgravescs) [inline docs/security.md]: secure yarn and hdfs with encryption.   Some companies don't consider this secure enough.  Not sure if it makes sense for us to say anything more though.

File: docs/security.md [hunk 3]
```diff
@@ -182,54 +580,70 @@ configure those ports.
...
-parameters, and <a href="{{site.SPARK_GITHUB_URL}}/tree/master/core/src/main/scala/org/apache/spark/SecurityManager.scala">
-<code>org.apache.spark.SecurityManager</code></a> for implementation details about security.
+## Long-Running Applications
+
+Long-running applications may run into issues if their run time exceeds the maximum delegation
+token lifetime configured in services it needs to access.
+
+Spark supports automatically creating new tokens for these applications when running in YARN mode.
+Kerberos credentials need to be provided to the Spark application via the `spark-submit` command,
+using the `--principal` and `--keytab` parameters.
```
Comment (tgravescs) [inline docs/security.md]: should we link the yarn page spark.yarn.keytab and spark.yarn.principal to this section as well

File: docs/running-on-yarn.md [hunk 4]
```diff
@@ -2,6 +2,8 @@
 layout: global
 title: Running Spark on YARN
 ---
+* This will become a table of contents (this text will be scraped).
```
[the remark this answers is not in the document]
Comment (vanzin) [reply docs/running-on-yarn.md]: This is the same wording as other pages, so I'd rather keep it the same. (And "scrape" is also correct.)

File: docs/security.md (hunk 3 above)
Comment (vanzin) [reply docs/security.md]: I moved the yarn/kerberos configs to a new separate table in the YARN page. That should make them more visible.

File: docs/running-on-yarn.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (tgravescs) [reply docs/running-on-yarn.md]: maybe I misread this, is it supposed to be this text will be scraped (meaning looked for in order to fill in TOC) or was it supposed to be scrapped (meaning thrown away)?

**Comment by tgravescs** [comment]
+1

File: docs/running-on-yarn.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (vanzin) [reply docs/running-on-yarn.md]: This text will be replaced with the TOC.

File: docs/security.md [hunk 5]
```diff
@@ -3,47 +3,291 @@ layout: global
...
+    <code>commons.crypto</code> prefix.
+  </td>
+</tr>
+<tr>
+  <td><code>spark.network.crypto.saslFallback</code></td>
+  <td>true</td>
+  <td>
+    Whether to fall back to SASL authentication if authentication fails using Spark's internal
+    mechanism. This is useful when the application is connecting to old shuffle services that
+    do not support the internal Spark authentication protocol. On the shuffle service side, enabling
```
Comment (squito) [inline docs/security.md]: is this part right -- enabling the option (not disabling) makes the shuffle service refuse to fallback to sasl?  from a quick scan I expected the opposite:

https://github.com/apache/spark/blob/master/common/network-common/src/main/java/org/apache/spark/network/crypto/AuthRpcHandler.java#L97

File: docs/security.md [hunk 6]
```diff
@@ -182,54 +582,70 @@ configure those ports.
...
+and HBase.
+
+When using a Hadoop filesystem (such HDFS or WebHDFS), Spark will acquire the relevant tokens
+for the service hosting the user's home directory.
+
+An HBase token will be obtained if HBase is in the application's classpath, and the HBase
+configuration has Kerberos authentication turned (`hbase.security.authentication=kerberos`).
+
+Similarly, a Hive token will be obtained if Hive is in the classpath, and the configuration includes
+a URIs for remote metastore services (`hive.metastore.uris` is not empty).
```
Comment (squito) [inline docs/security.md]: nit: either "a URI" or "URIs"

**Review by squito — COMMENTED** [review]
mostly fine, a couple typos, and perhaps a real doc bug on spark.network.crypto.saslFallback

**Comment by squito** [comment]
lgtm

**Comment by kiszk** [comment]
retest this please

**Comment by vanzin** [comment]
No more comments, so merging to master.

**Pull request closed without merging**